rdtools/soiling.py

```diff
@@ -0,0 +1,987 @@
+                    end = inter_start + row.run_loss
+                    inter_start = 1
+                results_rand['start_loss'] = start_list
+
+            else:
+                raise ValueError("Invalid method specification")
+
+            df_rand = df_rand.reset_index() \
+                             .merge(results_rand, how='left', on='run') \
+                             .set_index('date')
+            df_rand['loss'] = np.nan
+            df_rand['days_since_clean'] = \
+                (df_rand.index - df_rand.start).dt.days
+            df_rand['loss'] = df_rand.start_loss + \
+                df_rand.days_since_clean * df_rand.run_slope
+
+            df_rand['soil_insol'] = df_rand.loss * df_rand.insol
+
+            monte_losses.append(df_rand.soil_insol.sum() / df_rand.insol[~df_rand.soil_insol.isnull()].sum())
+            random_profile = df_rand['loss'].copy()
+            random_profile.name = 'stochastic_soiling_profile'
+            random_profiles.append(random_profile)
+
+        self.random_profiles = random_profiles
+        self.monte_losses = monte_losses
+
+    def run(self, reps=1000, day_scale=14, clean_threshold='infer',
+            trim=False, method='half_norm_clean',
+            clean_criterion='shift', precip_threshold=0.01, min_interval_length=2,
+            exceedance_prob=95.0, confidence_level=68.2, recenter=True,
+            max_relative_slope_error=500.0, max_negative_step=0.05):
+        '''
+        Run the SRR method from beginning to end.  Perform the stochastic rate
+        and recovery soiling loss calculation. Based on the methods presented
+        in Deceglie et al. JPV 8(2) p547 2018.
+
+        Parameters
+        ----------
+        reps : int, default 1000
+            number of Monte Carlo realizations to calculate
+        day_scale : int, default 14
+            The number of days to use in rolling median for cleaning detection,
+            and the maximum number of days of missing data to tolerate in a
+            valid interval
+        clean_threshold : float or 'infer', default 'infer'
+            The fractional positive shift in rolling median for cleaning
+            detection. Or specify 'infer' to automatically use outliers in the
+            shift as the threshold.
+        trim : bool, default False
+            Whether to trim (remove) the first and last soiling intervals to
+            avoid inclusion of partial intervals
+        method : str, default 'half_norm_clean'
+            How to treat the recovery of each cleaning event:
+
+            * 'random_clean' - a random recovery between 0-100%
+            * 'perfect_clean' - each cleaning event returns the performance
+              metric to 1
+            * 'half_norm_clean' (default) - The three-sigma lower bound of
+              recovery is inferred from the fit of the following interval, the
+              upper bound is 1 with the magnitude drawn from a half normal
+              centered at 1
```

Comment:
"half normal centered at 1" is this a truncated normal with upper bound at 1 and lower bound at the three-sigma value? If so, suggest "a random recovery sampled from a normal distributed with mean 1 and sigma = lower  bound / 3 (could use a name for this), truncated between 3*sigma and 1"

Reply:
It isn't technically truncated on the lower edge. Perhaps it should be truncated based on the end of the last period, but for now I'll try to be more precise: 

> The starting point of each interval is taken randomly from a half normal distribution with its mode at 1 and its three-sigma point at the intercept of the fit to the interval.

Reply:
I understand now. Probably should drop the phrase "lower bound" in other places where 'half_norm_clean' is described. Perhaps in a future PR the sampling could be truncated to avoid the very rare event that soiling increases after the cleaning event, due to an extreme sample resulting in a negative recovery value. 

"mode at 1" is unambiguous but (splitting hairs a bit) here sigma is a distribution parameter rather than the standard deviation of the folded normal. "three-sigma point" might be misunderstood as three standard deviations of the folded normal. Maybe more clear to describe the distribution in terms of it's parameters, mu = 1 and sigma equal to 1/3 of the intercept of the fit to the interval.

Reply:
Thanks, that's a good suggestion. It's actually 1/3 * (1-intercept), I'll update each mention.